src/dune/dune_env.ml

```diff
@@ -14,6 +14,13 @@ module Stanza = struct
     and+ cxx = Ordered_set_lang.Unexpanded.field "cxx_flags" ?check in
     Foreign.Language.Dict.make ~c ~cxx
 
+  let menhir_flags ~since =
+    let check =
+      Option.map since ~f:(fun since ->
+          Dune_lang.Syntax.since Stanza.syntax since)
```

Comment:
```suggestion
          Dune_lang.Syntax.since Menhir.syntax since)
```

So that the versioning is tied to `(using menhir 2.1)` rather than `(lang dune 2.2)`. You also need to add a 2.1 version to the menhir extension by replacing `(2, 0)` by `(2, 1)` in the definition of `Menhir.syntax`.

`Menhir.syntax` is defined inside `dune_file.ml` BTW

Reply:
Thanks for the hints, I changed to `menhir`.

Reply:
Unfortunately this introduces a dependency cycle, but I'm not sure how to resolve this.

Reply:
Can it be resolved by moving the definition in the file? Here `Menhir` should refer to the `Menhir` sub-module in `dune_file.ml` rather than the toplevel `Menhir` module

Reply:
I tried using `Dune_file.Menhir.syntax` instead and it did not work. I suspect that one needs to move `Menhir.syntax` into an own file like it is done for `Stanza.syntax`.

Reply:
Ah yes, of course. I thought we were in `dune_file.ml` for some reason. Moving `Menhir.syntax` outside of `dune_file.ml` seems right. In fact, we can move the whole `Menhir` sub-module to its own file. I'm wondering if we can't just move that code to the toplevel `Menhir` module actually :thinking:

Reply:
Moving the `Menhir` stuff into an own file does not really help, also moving `syntax` into `menhir.ml` does work.

Reply:
In both cases I still got the cyclic dependency.

Reply:
I had a look and I see why. Ideally we would make `Dune_env` extensible in a generic way so that its code doesn't mention menhir at all. For the time being, you can indeed just move only the `syntax` value to a dedicated file.

Reply:
I moved the definition into a new file `menhir_stanza` and changed the usages to use the correct one.